ios-base/Navigators/Navigator.swift

```diff
@@ -0,0 +1,215 @@
+   - animated: Animate the transition or not.
+   - completion: Completion handler.
+   */
+  func navigate(to router: Navigator, animated: Bool, completion: (() -> Void)?)
+
+  /**
+   Handles backwards navigation through the stack.
+   */
+  func pop(animated: Bool)
+
+  /**
+   Handles backwards navigation through the stack.
+   - Parameters:
+   - route: The index of the route of your navigation action.
+   - animated: Animate the transition or not.
+   */
+  func popTo(index: Int, animated: Bool)
+
+  /**
+   Handles backwards navigation through the stack.
+   - Parameters:
+   - route: The destination route of your navigation action.
+   - animated: Animate the transition or not.
+   */
+  func popTo(route: Route, animated: Bool)
+
+  /**
+   Dismiss your current ViewController.
+   - Parameters:
+   - animated: Animate the transition or not.
+   - completion: Completion handler.
+   */
+  func dismiss(animated: Bool, completion: (() -> Void)?)
+}
+
+public extension Navigator {
+  func navigate(to route: Route, with transition: TransitionType, animated: Bool = true, completion: (() -> Void)? = nil) {
+    let viewController = route.screen
+    switch transition {
+    case .modal:
+      route.transitionConfigurator?(currentViewController, viewController)
+      currentViewController?.present(viewController, animated: animated, completion: completion)
+    case .push:
+      route.transitionConfigurator?(currentViewController, viewController)
+      rootViewController?.pushViewController(viewController, animated: animated)
+    case .reset:
+      route.transitionConfigurator?(nil, viewController)
+      rootViewController?.setViewControllers([viewController], animated: animated)
+    case .changeRoot:
+      UIView.animate(withDuration: 0.3) { [weak self] in
+        let navigation = viewController.embedInNavigationController()
+        UIApplication.shared.keyWindow?.rootViewController = navigation
+        self?.rootViewController = navigation
+      }
+    }
+  }
+
+  func navigate(to router: Navigator, animated: Bool, completion: (() -> Void)?) {
+    guard let viewController = router.rootViewController else {
+      assert(false, "Navigator does not have a root view controller")
+      return
+    }
+
+    currentViewController?.present(viewController, animated: animated, completion: completion)
+  }
+
+  func pop(animated: Bool = true) {
+    rootViewController?.popViewController(animated: animated)
+  }
+
+  func popToRoot(animated: Bool = true) {
+    rootViewController?.popToRootViewController(animated: animated)
+  }
+
+  func popTo(index: Int, animated: Bool = true) {
+    guard
+      let viewControllers = rootViewController?.viewControllers,
+      viewControllers.count > index
+    else { return }
+    let viewController = viewControllers[index]
+    rootViewController?.popToViewController(viewController, animated: animated)
+  }
+
+  func popTo(route: Route, animated: Bool = true) {
+    guard
```

Comment:
I like this guard indentation, @glm4 what do you think about adding this to our style guides ?